src/Meziantou.Analyzer/Rules/ReturnTaskDirectlyAnalyzer.cs

```diff
@@ -0,0 +1,215 @@
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+using Meziantou.Analyzer.Internals;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Meziantou.Analyzer.Rules;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ReturnTaskDirectlyAnalyzer : DiagnosticAnalyzer
+{
+    private static readonly DiagnosticDescriptor s_rule = new(
+        RuleIdentifiers.ReturnTaskDirectly,
+        title: "Return Task instead of awaiting it",
+        messageFormat: "Return Task instead of awaiting it",
+        RuleCategories.Performance,
+        DiagnosticSeverity.Info,
+        isEnabledByDefault: true,
+        description: "Returning a Task directly instead of awaiting it increases performance, as the compiler does not need to create an additional state machine.",
+        helpLinkUri: RuleIdentifiers.GetHelpUri(RuleIdentifiers.ReturnTaskDirectly));
+
+    public override void Initialize(AnalysisContext context)
+	{
+		context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+		context.EnableConcurrentExecution();
+		context.RegisterCompilationStartAction(OnCompilationStart);
+	}
+
+	private static void OnCompilationStart(CompilationStartAnalysisContext context)
+	{
+		context.RegisterSyntaxNodeAction(OnMethodAnalysis, SyntaxKind.MethodDeclaration, SyntaxKind.LocalFunctionStatement, SyntaxKind.ParenthesizedLambdaExpression);
```

Comment:
Have you tried using the Operation analysis instead of SyntaxNode analysis? Most of the time, this simplifies the code as you don't have to deal with the different ways to declare the same concept. Also, you don't need to use the SemanticModel as the main info as already accessible from operations.

Reply:
Yeah, I was thinking that as well as I kept adding cases to the analyzer. I'll see if I can find an elegant solution.

Reply:
Hmm, this doesn't seem as straightforward as I thought. Access to method modifiers and return types is easier with a `SyntaxNode` and also there doesn't seem to be an `IOperation` for a lambda expression, so I would need to process that as a `SyntaxNode` anyway.

Reply:
> there doesn't seem to be an IOperation for a lambda expression

You can use `IDelegateCreationOperation` or `OperationKind.DelegateCreation`

> Access to method modifiers and return types is easier with a `SyntaxNode`

I don't think you need to access modifiers. Your code already use properties from IMethodSymbol.


````c#
context.RegisterOperationBlockAction(ctx => { var block = ctx.OperationBlocks; var method = ctx.OwningSymbol; });
context.RegisterOperationAction(ctx => { var op = ctx.Operation; }, OperationKind.LocalFunction);
context.RegisterOperationAction(ctx => { var op = ctx.Operation; }, OperationKind.DelegateCreation);
````